File: android/app/src/main/java/com/adasa/foodplan/domain/model/ShoppingList.kt
package com.adasa.foodplan.domain.model

data class ShoppingList(
    val period: ShoppingPeriod,
    val categories: List<ShoppingCategory>
) {
    val totalItems: Int get() = categories.sumOf { it.items.size }
}

data class ShoppingPeriod(
    val startDate: kotlinx.datetime.LocalDate,
    val endDate:   kotlinx.datetime.LocalDate,
    /** All recipes in this period, for the filter selector */
    val recipes: List<SelectableRecipe>
) {
    val recipeNames: List<String> get() = recipes.map { it.name }
}

/** A recipe that can be toggled in/out of the shopping list */
data class SelectableRecipe(
    val id:   String,
    val name: String
)

data class ShoppingCategory(
    val name:  String,
    val emoji: String,
    val items: List<ShoppingItem>
)

/** Per-recipe breakdown: how many grams of this ingredient does recipe X need */
data class RecipeContribution(
    val recipeId:   String,
    val recipeName: String,
    val grams:      Double
)

data class ShoppingItem(
    val ingredientId:    String,
    val name:            String,
    val totalGrams:      Double,
    val unit:            ShoppingUnit,
    /** Non-empty: each recipe that uses this ingredient + how many grams */
    val contributions:   List<RecipeContribution>
) {
    /** Convenience — recipe names only, for legacy use */
    val usedInRecipes: List<String> get() = contributions.map { it.recipeName }.distinct().sorted()
}

enum class ShoppingUnit { GRAMS, PIECES, DECILITERS }

File: flutter-app/lib/domain/model/shopping_list.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'shopping_list.freezed.dart';

// ── Enums ─────────────────────────────────────────────────────────────────────

enum ShoppingUnit { grams, pieces, deciliters }

// ── Domain models ─────────────────────────────────────────────────────────────

@freezed
class ShoppingList with _$ShoppingList {
  const factory ShoppingList({
    required ShoppingPeriod period,
    required List<ShoppingCategory> categories,
  }) = _ShoppingList;

  const ShoppingList._();

  int get totalItems => categories.fold(0, (sum, c) => sum + c.items.length);
}

@freezed
class ShoppingPeriod with _$ShoppingPeriod {
  const factory ShoppingPeriod({
    required DateTime startDate,
    required DateTime endDate,
    @Default(<SelectableRecipe>[]) List<SelectableRecipe> recipes,
  }) = _ShoppingPeriod;

  const ShoppingPeriod._();

  List<String> get recipeNames => recipes.map((r) => r.recipeName).toList();
}

@freezed
class SelectableRecipe with _$SelectableRecipe {
  const factory SelectableRecipe({
    required String recipeId,
    required String recipeName,
    @Default(true) bool isSelected,
  }) = _SelectableRecipe;
}

@freezed
class ShoppingCategory with _$ShoppingCategory {
  const factory ShoppingCategory({
    required String name,
    required String emoji,
    required List<ShoppingItem> items,
  }) = _ShoppingCategory;
}

@freezed
class ShoppingItem with _$ShoppingItem {
  const factory ShoppingItem({
    required String ingredientId,
    required String name,
    required double totalGrams,
    required ShoppingUnit unit,
    required List<RecipeContribution> contributions,
  }) = _ShoppingItem;

  const ShoppingItem._();

  List<String> get usedInRecipes =>
      contributions.map((c) => c.recipeName).toSet().toList()..sort();
}

@freezed
class RecipeContribution with _$RecipeContribution {
  const factory RecipeContribution({
    required String recipeName,
    required double grams,
  }) = _RecipeContribution;
}
